--- old/Parser/grammar.c	2010-05-09 18:52:27.000000000 +0300
+++ new/Parser/grammar.c	2011-07-11 11:45:00.000000000 +0300
@@ -9,6 +9,9 @@
 #include "token.h"
 #include "grammar.h"
 
+#include <wctype.h>
+extern int Py_USELOCALE;
+
 extern int Py_DebugFlag;
 
 grammar *
@@ -179,7 +182,15 @@
     }
 
     if (lb->lb_type == STRING) {
-        if (isalpha(Py_CHARMASK(lb->lb_str[1])) ||
+        int alpha;
+        if (Py_USELOCALE) {
+            wchar_t x=0;
+            mbtowc( &x, lb->lb_str+1, strlen( lb->lb_str+1) );
+            alpha = iswalpha( x);
+        } else
+            alpha = isalpha(Py_CHARMASK(lb->lb_str[1]));
+
+        if (alpha ||
             lb->lb_str[1] == '_') {
             char *p;
             char *src;
